### DISCUSSION TITLE
Efficient approach for moving tree nodes (up/down) with children in slickgrid-universal

### DISCUSSION THREAD

**Discussion by hwlv** [ghiscoding/slickgrid-universal#1754] [Ideas]
Looking for an efficient way to implement up/down movement of nodes in a tree grid where moving a parent node should also move all its children while maintaining the hierarchy.
Current Challenge:
When clicking up/down buttons to move nodes:

Parent node movement should include all children
Keep hierarchy intact
Handle proper node reordering
Maintain correct indentation levels

Current approach:

```js
class TreeGridMovement {
  moveNode(direction: 'up' | 'down') {
    const selectedNode = this.getSelectedNode();
    const allNodes = this.grid.getData();
    const currentIndex = allNodes.indexOf(selectedNode);
    
    // Get all children nodes
    const subtree = this.getAllDescendants(selectedNode);
    const subtreeSize = subtree.length;
    
    // Calculate target index
    const targetIndex = direction === 'up' 
      ? Math.max(0, currentIndex - 1)
      : Math.min(allNodes.length - subtreeSize, currentIndex + 1);
      
    // Move entire subtree
    const nodesToMove = subtree;
    nodesToMove.forEach((node, i) => {
      const idx = allNodes.indexOf(node);
      allNodes.splice(idx, 1);
      allNodes.splice(targetIndex + i, 0, node);
    });
    
    this.updateLevels(allNodes);
    this.grid.invalidate();
  }
}
    ```
Questions:

What's the most efficient way to handle batch node movement?
Are there better approaches to maintain tree structure during moves?
Any suggestions for optimizing level recalculation?

Looking forward to your insights on improving this implementation.

**Comment by ghiscoding** [comment]
The approach is already shown in the Example 5, you need to update your row detail and `parentId` and then reassign to the `datasetHierarchical` (or reassign to the regular `dataset`, that should also trigger an update).

https://github.com/ghiscoding/slickgrid-universal/blob/8efae0e448aa8a52c380d640e5838918584f7fd3/examples/vite-demo-vanilla-bundle/src/examples/example06.ts#L281-L298

A reminder again that SlickGrid by default doesn't support Tree Data natively, what I did to make it work is to keep 2 data arrays in sync, the 2 arrays are the flat dataset and the tree dataset (hierarchical), when one changes then the other one needs to be resync. When you change one of the array, the other should resync automatically. On a very large dataset, this might not be the most efficient, but it is really the only way to get it working in SlickGrid

Another option is to call `invalidateHierarchicalDataset()` from the Grid Service, invalidating the hierarchical data will force a refresh in the grid with latest data

https://github.com/ghiscoding/slickgrid-universal/blob/8efae0e448aa8a52c380d640e5838918584f7fd3/packages/common/src/services/grid.service.ts#L864-L870
